WT-12827 Add script to install tcmalloc on workstations

marcbutler opened the pull request:
The equivalent of "test/evergreen/tcmalloc_install_or_build.sh" for developer workstations.

If successful the resulting directory TCMALLOC_LIB in the current git workspace will contained a usable shared object, readme and helper scripts for executing and debugging with libtcmalloc.so

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,72 @@
+
+set -euf -o pipefail
+
+HERE=$(git rev-parse --show-toplevel) || echo "FATAL Run in root of git workspace"
```
`set -e` doesn't have an effect on this.
If `git rev-parse` fails, then the exit status of the whole command will be the `echo`'s one which is always "success", therefore the script will continue execution even if `rev-parse` fails and the error is reported.

A good practice is to add this function:
```bash
die() { echo "$@"; exit 1; }
```

and use it instead of `echo` like this:
```bash
command || die "error"
```

Same applies to other error checks.

marcbutler replied:
Done

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,72 @@
+which bazel || echo "FATAL Build tool bazel not found in path"
+
+# Use mongo toolchain
+[[ -d /opt/mongodbtoolchain/v4/bin ]] || "FATAL Require mongodb toolchain."
```
missing 'echo' or 'die' before the error message

ershov reviewed (commented):
This is really cool, thanks for this tool!

This is my first sight, I'll give it another go later today.

lukech commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+#
+# test/evergreen/tcmalloc_install_or_build.sh
+#
+# and the tcmalloc source version referenced here should be updated in
+# sync with that script.
```
Can we have the other script refer to this script as well? Or have the tcmalloc source version (lines #22--24, #41--44 below) abstracted into a common file that both scripts can source (to avoid potential version upgrade inconsistency)

marcbutler replied:
There isn't a logical place afaict for this given the different directories they live are at the top-level.
This is a rarely used script: so maybe it is better to require an argument that requires the user grab the relevant information from the script under `test/`.

lukech replied:
> maybe it is better to require an argument that requires the user grab the relevant information from the script under test/?

Sounds good. As long as users get some reminder about the tcmalloc version that should be installed (the version that's actively tested in Evergreen) while running this script.

lukech commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+[[ -f ${HERE}/CMakeLists.txt ]] || die "Does not look like a WT workspace - no CMakeLists.txt"
+[[ -d ${HERE}/TCMALLOC_LIB ]] && die "TCMALLOC_LIB already present"
+
+# Before building look for a prebuilt copy cached on this system.
+SRC_ID=mongo-SERVER-85737
+CACHED_DIR=${HOME}/.cache/wiredtiger/tcmalloc/${SRC_ID}
+CACHED_BIN=${HOME}/.cache/wiredtiger/tcmalloc/${SRC_ID}/TCMALLOC_LIB.tgz
+if [[ -r ${CACHED_BIN} ]]; then
+    echo "Installing cached copy from $CACHED_BIN"
+    tar zxf $CACHED_BIN
+    exit
+fi
```
I try to think about scenarios where this script needs to be run:

1. When a new workstation is spawned from Evergreen. As no prebuilt copy is in place, this code section will be a "no-op" when running this script. 
2. When a newer tcmalloc version becomes available to which we hope to upgrade. There's a prebuilt copy in place while outdated. Running this script would extract the tarball with a dated version and exit (on line #28). 

Do we expect users to be aware of the version upgrade and edit line #22 for the 2nd scenario?

marcbutler replied:
This functionality has been removed.

lukech commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+which bazel || echo "FATAL Build tool bazel not found in path"
+
+# Use mongo toolchain
+[[ -d /opt/mongodbtoolchain/v4/bin ]] || die "FATAL Requires mongodb toolchain."
```
Is mongodb toolchain a must or a nice-to-have for building tcmalloc? I'm wondering if it's possible to use this script on a local laptop where toolchain is not readily available (or available but in a directory other than the standard one).

marcbutler replied:
This script is intended to be for use by mongodb employed developers. If community developers want to work with tcmalloc they can load it from their respective environments.

lukech replied:
Thanks for clarifying it. I was actually thinking about internal users who use their laptops as the main development environment. Unfortunately, mongodb toolchain does not support macOS AFAIK. So they need to adjust this script before using it (which is fine).

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+
+HERE=$(git rev-parse --show-toplevel) || die "FATAL Run in root of git workspace"
+[[ -f ${HERE}/CMakeLists.txt ]] || die "Does not look like a WT workspace - no CMakeLists.txt"
+[[ -d ${HERE}/TCMALLOC_LIB ]] && die "TCMALLOC_LIB already present"
```
This probably should just `exit 0` as the purpose of the script is to make tcmalloc available. If it's already there, nothing to do then.

marcbutler replied:
It's a fair point, but this is a developer run script, and the developer thought they need to build tcmalloc: and the script is no repurposed to build tcmalloc only and can't so it's still arguably an error, and not likely to upset the developer.

marcbutler replied:
Given the test is simplistic in its test I think it is best to warn the developer nothing was done.

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+echo "No cached copy on this system. Attempting to build."
+
+# Requires bazel.
+which bazel || echo "FATAL Build tool bazel not found in path"
```
Should this be `die`, not `echo`?

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+
+# This is tcmalloc upstream revision 093ba93 source patched by the SERVER team.
+# https://github.com/mongodb-forks/tcmalloc/releases/tag/mongo-SERVER-85737
+PATCHED_SRC=mongo-SERVER-85737
```
Is `PATCHED_SRC` a duplicate of `SRC_ID` ?

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+PATCHED_TGZ_URL="https://github.com/mongodb-forks/tcmalloc/archive/refs/tags/${PATCHED_TGZ}"
+PATCHED_SRC_DIR="tcmalloc-${PATCHED_SRC}"
+
+curl --retry 5 -L $PATCHED_TGZ_URL -sS --max-time 120 --fail --output ${PATCHED_TGZ}
```
Nice set of flags for `curl`!

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+cp $PATCHED_SRC_DIR/bazel-bin/libtcmalloc.so TCMALLOC_LIB
+
+# Generate script to define "with_tcmalloc" convenience function. 
+cat <<EOF > TCMALLOC_LIB/with_tcmalloc.sh
```
optional improvement suggestion: if you quote the termination string, there's no variable substitution in the "heredoc" so you don't have to quote each `$`:
```bash
$ cat << 'EOF'
> $PATH
> EOF
$PATH
```

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+
+HERE=$(git rev-parse --show-toplevel) || die "FATAL Run in root of git workspace"
+[[ -f ${HERE}/CMakeLists.txt ]] || die "Does not look like a WT workspace - no CMakeLists.txt"
+[[ -d ${HERE}/TCMALLOC_LIB ]] && die "TCMALLOC_LIB already present"
```
Maybe it's better to check for the presence of `TCMALLOC_LIB/libtcmalloc.so`, not just directory. Build could fail or be aborted.

marcbutler replied:
I'd rather that the user be forced to verify and delete the directory, rather than try to be too precise, as this script generates other files.
If we are running this script *a lot* something has gone terrribly wrong.

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+
+export LD_PRELOAD=$PWD/TCMALLOC_LIB/libtcmalloc.so
+
+This will be ineffect until you exit the current shell.
```
```suggestion
This will be in effect until you exit the current shell.
```

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+TOPDIR=\$(git rev-parse --show-toplevel) || (echo "FATAL Not a git repo" && return)
+SOPATH=\$TOPDIR/TCMALLOC_LIB/libtcmalloc.so 
+[[ -f \$SOPATH ]] || echo "FATAL libtcmalloc.so not found"
+eval "with_tcmalloc() { LD_PRELOAD=\$SOPATH \\$* ;}"
```
```suggestion
eval "with_tcmalloc() { LD_PRELOAD=\$SOPATH "\$@" ;}"
```

marcbutler replied:
What about the nested double quotes?

ershov commented on tools/install-tcmalloc.sh:
```diff
@@ -0,0 +1,142 @@
+# Generate script to define "with_tcmalloc" convenience function. 
+cat <<EOF > TCMALLOC_LIB/with_tcmalloc.sh
+with_tcmalloc() {
+TOPDIR=\$(git rev-parse --show-toplevel) || (echo "FATAL Not a git repo" && return)
```
```suggestion
TOPDIR=\$(git rev-parse --show-toplevel) || { echo "FATAL: Not a git repo"; return 1; }
```

marcbutler commented:
I've reduced the purpose of the script, and have outlined that purpose in the prologue comments at the top of the script. I've also addressed some issues, some I don't understand.
Yes ideally they would use the same source code identifier but the rework required, makes me think it is better spun off into a new-eng ticket if we are adamant about it. This is a stop gap script, and over-engineering it I think is not a good investment of effort.

marcbutler commented on tools/build-tcmalloc.sh:
```diff
@@ -0,0 +1,136 @@
+cp $PATCHED_SRC_DIR/bazel-bin/libtcmalloc.so TCMALLOC_LIB
+
+# Generate script to define "with_tcmalloc" convenience function. 
+cat << 'EOF' > TCMALLOC_LIB/with_tcmalloc.sh
+with_tcmalloc() {
+TOPDIR=\$(git rev-parse --show-toplevel) || { echo "FATAL Not a git repo" && return
+SOPATH=\$TOPDIR/TCMALLOC_LIB/libtcmalloc.so 
+[[ -f \$SOPATH ]] || echo "FATAL libtcmalloc.so not found"
+eval "with_tcmalloc() { LD_PRELOAD=\$SOPATH \\$* ;}"
+unset TOPDIR SOPATH
+}
+EOF
```
Sorry I don't follow. It is still being generated in the example you provide which has syntax errors.

ajmorton commented on tools/build-tcmalloc.sh:
```diff
@@ -0,0 +1,136 @@
+[[ -d ${HERE}/TCMALLOC_LIB ]] && die "TCMALLOC_LIB already present"
+
+# Requires bazel.
+which bazel || echo "FATAL Build tool bazel not found in path"
```
```suggestion
which bazel || die "FATAL Build tool bazel not found in path"
```

ajmorton commented on tools/build-tcmalloc.sh:
```diff
@@ -0,0 +1,136 @@
+# Generate script to define "with_tcmalloc" convenience function. 
+cat << 'EOF' > TCMALLOC_LIB/with_tcmalloc.sh
+with_tcmalloc() {
+TOPDIR=\$(git rev-parse --show-toplevel) || { echo "FATAL Not a git repo" && return
+SOPATH=\$TOPDIR/TCMALLOC_LIB/libtcmalloc.so 
+[[ -f \$SOPATH ]] || echo "FATAL libtcmalloc.so not found"
+eval "with_tcmalloc() { LD_PRELOAD=\$SOPATH \\$* ;}"
+unset TOPDIR SOPATH
```
How do these lines work? I'd assume `echo "FATAL Not a git repo"` always succeeds so we `return` and never reach the subsequent `SOPATH=\$TOPDIR/TCMALLOC_LIB/libtcmalloc.so ` line

ajmorton reviewed (approved):
LGTM 👍

marcbutler commented:
I'd like to close this out, and I know I have your okay -- but this significantly reworks the script so I felt I should ask for a re-review.

ajmorton commented on tools/tcmalloc/define-with_tcmalloc.sh:
```diff
@@ -0,0 +1,17 @@
+    return
+fi
+
+eval "with_tcmalloc() { LD_PRELOAD=$SO__ \"\$@\"; }"
```
```suggestion
eval "with_tcmalloc() { LD_PRELOAD=$SO__:$LD_PRELOAD \"\$@\"; }"
```

ajmorton reviewed (approved):
LGTM 👍

lukech commented on tools/tcmalloc/README.txt:
```diff
@@ -0,0 +1,45 @@
+Explicit
+========
+
+Alternatively you can source the script with_tcmalloc.sh in this
```
It looks like there's only a `define-with_tcmalloc.sh`, not `with_tcmalloc.sh` script in the directory.

marcbutler replied:
So the script defines the bash function `with_tcmalloc` so that you can run:
```
$ with_tcmalloc test/csuite/mytest
```

lukech reviewed (approved):
LGTM % nit
